rfq/negotiator.go

```diff
@@ -445,6 +465,101 @@ func (n *Negotiator) HandleOutgoingSellOrder(order SellOrder) {
+	// Convert the expiry timestamp into a time.Time.
+	actualExpiry := time.Unix(int64(expiryUnixTimestamp), 0)
+	diff := actualExpiry.Unix() - time.Now().Unix()
+	return diff >= minRateTickExpiryLifetime
+}
+
+// priceWithinBounds returns true if the difference between the first price and
+// the second price is within the acceptable bounds.
+func (n *Negotiator) pricesWithinBounds(firstPrice lnwire.MilliSatoshi,
+	secondPrice lnwire.MilliSatoshi) bool {
+
+	deviation := 10000 * math.Abs(float64(secondPrice-firstPrice)) /
+		float64(secondPrice)
+	return deviation <= float64(n.cfg.AcceptPriceDeviationBp)
+}
+
+// HandleIncomingBuyAccept handles an incoming buy accept message. This method
+// is called when a peer accepts a quote request from this node. The method
+// checks the price and expiry time of the quote accept message. If the quote is
+// acceptable, the method calls the finalise callback function.
+func (n *Negotiator) HandleIncomingBuyAccept(msg rfqmsg.BuyAccept,
+	finalise func(rfqmsg.BuyAccept)) error {
+
+	// Query the price oracle asynchronously using a separate goroutine.
+	n.Wg.Add(1)
+	go func() {
+		defer n.Wg.Done()
+
+		// Ensure that the quote expiry time is within acceptable
+		// bounds.
+		//
+		// TODO(ffranr): Sanity check the buy accept quote expiry
+		//  timestamp given the expiry timestamp provided by the price
+		//  oracle.
+		if !n.expiryWithinBounds(msg.Expiry) {
+			// The expiry time is not within the acceptable bounds.
+			// We will return without calling the quote accept
+			// callback.
+			log.Debugf("buy accept quote expiry time is not "+
+				"within acceptable bounds (expiry=%d)",
+				msg.Expiry)
+			return
+		}
+
+		if n.cfg.SkipAcceptQuotePriceCheck {
+			// Skip the price check.
+			finalise(msg)
+			return
+		}
+
+		// The buy accept message contains an ask price. This price
+		// is the price that the peer is willing to accept in order to
+		// sell the asset that we are buying.
+		//
+		// We will sanity check that price by querying our price oracle
+		// for an ask price. We will then compare the ask price returned
+		// by the price oracle with the ask price provided by the peer.
+		oraclePrice, _, err := n.queryAskFromPriceOracle(
+			&msg.Peer, msg.AssetID, nil,
+			msg.AssetAmount, nil,
+		)
+		if err != nil {
+			// The price oracle returned an error. We will return
+			// without calling the quote accept callback.
+			err = fmt.Errorf("negotiator failed to query price "+
+				"oracle when handling incoming buy accept "+
+				"message: %w", err)
+			log.Errorf("Error calling price oracle: %v", err)
+			n.cfg.ErrChan <- err
+
+			return
+		}
+
+		// Ensure that the peer provided price is reasonable given the
+		// price provided by the price oracle service.
+		if !n.pricesWithinBounds(msg.AskPrice, oraclePrice) {
+			// The price is not within the acceptable tolerance.
+			// We will return without calling the quote accept
+			// callback.
+			log.Debugf("buy accept price is not within "+
```

Comment:
I wouldn't think of it as an error if the price is unacceptable. Perhaps it should lead to an event being published by the RFQ manager to its subscribers?

Reply:
The problem is that when the user runs an asset channel related command (for example `litcli ln addinvoice` or `litcli ln payinvoice`), the code calls into the RFQ service to create a buy/sell order. That code then waits for either an accept or error to arrive. If nothing happens, because we simply drop/ignore the accepted quote, then the command will time out and the user needs to check the log to find out why.
So I do think we should return an error somehow. Maybe we can differentiate between an actual hard error (something didn't work on the protocol level) and a soft error (something didn't match my expectations).

Reply:
Yes, I agree with what you've said. That's why returning an event would be appropriate. I think the RPC endpoint just blocks waiting for an event anyway. I could be wrong, will look into it.